hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -523,15 +523,15 @@ void updateCurrBlockRef(HFileBlock block) {
       if (block != null && curBlock != null && block.getOffset() == curBlock.getOffset()) {
         return;
       }
-      if (this.curBlock != null) {
+      if (this.curBlock != null && !this.curBlock.isOnHeap()) {
```

Comment:
We still have this but in a different way. Instead of MemType the block knows its type.

Reply:
Discussed with @anoopsjohn before,  Maitaining a MemType inside the block will make the HFileBlock complex and easy to write a bug, because we need to handle the shared & non-shared logics  at the same paths or methods.
Separating them into two classes will be more clear.

Reply:
Ya its ok we dont keep the memtype in single class type. But why we can not create 2 block types SharedMem vs ExclusiveMem rather than on heap vs off heap?

Reply:
Addressed this in the new patch.  Thanks.